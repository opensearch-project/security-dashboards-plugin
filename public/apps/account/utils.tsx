/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { API_AUTH_LOGOUT, OPENID_AUTH_LOGOUT, SAML_AUTH_LOGOUT } from '../../../common';
import { API_ENDPOINT_ACCOUNT_INFO } from './constants';
import { AccountInfo } from './types';
import { httpGet, httpGetWithIgnores, httpPost } from '../configuration/utils/request-utils';
import { setShouldShowTenantPopup } from '../../utils/storage-utils';

export function fetchAccountInfo(http: HttpStart): Promise<AccountInfo> {
  return httpGet(http, API_ENDPOINT_ACCOUNT_INFO);
}

export async function fetchAccountInfoSafe(http: HttpStart): Promise<AccountInfo | undefined> {
  return httpGetWithIgnores<AccountInfo>(http, API_ENDPOINT_ACCOUNT_INFO, [401]);
}

export async function logout(http: HttpStart, logoutUrl?: string): Promise<void> {
  await httpPost(http, API_AUTH_LOGOUT);
  setShouldShowTenantPopup(null);
  // Clear everything in the sessionStorage since they can contain sensitive information
  sessionStorage.clear();
  // When no basepath is set, we can take '/' as the basepath.
  const basePath = http.basePath.serverBasePath ? http.basePath.serverBasePath : '/';
  const nextUrl = encodeURIComponent(basePath);
  window.location.href =
    logoutUrl || `${http.basePath.serverBasePath}/app/login?nextUrl=${nextUrl}`;
}

export async function samlLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  window.location.href = `${http.basePath.serverBasePath}${SAML_AUTH_LOGOUT}`;
}

export async function openidLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${OPENID_AUTH_LOGOUT}`;
}

export async function externalLogout(http: HttpStart, logoutEndpoint: string): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${logoutEndpoint}`;
}

export async function updateNewPassword(
  http: HttpStart,
  newPassword: string,
  currentPassword: string
): Promise<void> {
  await httpPost(http, API_ENDPOINT_ACCOUNT_INFO, {
    password: newPassword,
    current_password: currentPassword,
  });
}
