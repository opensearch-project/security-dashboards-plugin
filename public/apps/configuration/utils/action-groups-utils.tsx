/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { map } from 'lodash';
import { API_ENDPOINT_ACTIONGROUPS, CLUSTER_PERMISSIONS, INDEX_PERMISSIONS } from '../constants';
import { DataObject, ActionGroupItem, ActionGroupUpdate, ObjectsMessage } from '../types';
import { httpDelete, httpGet, httpPost } from './request-utils';
import { getResourceUrl } from './resource-utils';

export interface PermissionListingItem {
  name: string;
  type: 'Action group' | 'Single permission';
  reserved: boolean;
  allowedActions: string[];
  hasClusterPermission: boolean;
  hasIndexPermission: boolean;
}

export async function fetchActionGroups(http: HttpStart): Promise<DataObject<ActionGroupItem>> {
  const actiongroups = await httpGet<ObjectsMessage<ActionGroupItem>>({
    http,
    url: API_ENDPOINT_ACTIONGROUPS,
  });
  return actiongroups.data;
}

export function transformActionGroupsToListingFormat(
  rawData: DataObject<ActionGroupItem>
): PermissionListingItem[] {
  return map(rawData, (value: ActionGroupItem, key?: string) => ({
    name: key || '',
    type: 'Action group',
    reserved: value.reserved,
    allowedActions: value.allowed_actions,
    hasClusterPermission: value.type === 'cluster' || value.type === 'all',
    hasIndexPermission: value.type === 'index' || value.type === 'all',
  }));
}

export async function fetchActionGroupListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(await fetchActionGroups(http));
}

function getClusterSinglePermissions(): PermissionListingItem[] {
  return CLUSTER_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: true,
    hasIndexPermission: false,
  }));
}

function getIndexSinglePermissions(): PermissionListingItem[] {
  return INDEX_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: false,
    hasIndexPermission: true,
  }));
}

export async function getAllPermissionsListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return mergeAllPermissions(await fetchActionGroups(http));
}

export async function mergeAllPermissions(
  actionGroups: DataObject<ActionGroupItem>
): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(actionGroups)
    .concat(getClusterSinglePermissions())
    .concat(getIndexSinglePermissions());
}

export async function updateActionGroup(
  http: HttpStart,
  groupName: string,
  updateObject: ActionGroupUpdate
): Promise<ActionGroupUpdate> {
  return await httpPost({
    http,
    url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, groupName),
    body: updateObject,
  });
}

export async function requestDeleteActionGroups(http: HttpStart, groups: string[]) {
  for (const group of groups) {
    await httpDelete({ http, url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, group) });
  }
}
