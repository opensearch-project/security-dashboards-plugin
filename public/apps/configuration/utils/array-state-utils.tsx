/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { get, set, curry, PropertyName } from 'lodash';
import { Dispatch, SetStateAction } from 'react';

function resolveValue<T>(value: T | (() => T)) {
  return typeof value === 'function' ? (value as Function)() : value;
}

/**
 * Update an element in an sub array
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern"
 * @param newValue value to be updated
 *
 * e.g.
 *  currentState = [{a: 1}, {a: 2}]
 *  path = [0, 'a']
 *  newValue = 10
 * The newState will be [{a: 10}, {a: 2}]
 */
export function updateElementInArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  setStateCallback((prevState) => {
    const newState = [...prevState];
    set(newState, path, resolveValue(newValue));
    return newState;
  });
}

export const updateElementInArrayHandler = curry(updateElementInArray);

/**
 * Append an element to an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param newValue value to be updated
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  newValue = 3
 * The newState will be [1, 2, 3]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  newValue = 5
 * The newState will be [{a: [1, 2, 5]}, {a: [3, 4]}]
 */
export function appendElementToArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  const resolvedNewValue = resolveValue(newValue);
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      return [...prevState, resolvedNewValue];
    } else {
      const newArray = [...(get(prevState, path) as T[]), resolvedNewValue];
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}

/**
 * Remove an element from an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param index index of element to be removed
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  index = 0
 * The newState will be [2]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  index = 1
 * The newState will be [{a: [1]}, {a: [3, 4]}]
 */
export function removeElementFromArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  index: number
) {
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      const newState = [...prevState];
      newState.splice(index, 1);
      return newState;
    } else {
      const newArray = [...(get(prevState, path) as T[])];
      newArray.splice(index, 1);
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}
