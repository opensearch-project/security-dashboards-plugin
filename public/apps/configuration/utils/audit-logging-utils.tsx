/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { AuditLoggingSettings } from '../panels/audit-logging/types';
import { API_ENDPOINT_AUDITLOGGING, API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import { httpGet, httpPost } from './request-utils';

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await httpPost({ http, url: API_ENDPOINT_AUDITLOGGING_UPDATE, body: updateObject });
}

export async function getAuditLogging(http: HttpStart): Promise<AuditLoggingSettings> {
  const rawConfiguration = await httpGet<any>({ http, url: API_ENDPOINT_AUDITLOGGING });
  return rawConfiguration?.config;
}
