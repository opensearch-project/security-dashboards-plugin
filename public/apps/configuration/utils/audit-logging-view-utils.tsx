/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */
import {
  EuiDescribedFormGroup,
  EuiFlexGrid,
  EuiForm,
  EuiFormRow,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiSwitch,
  EuiText,
  EuiTitle,
} from '@elastic/eui';

import React from 'react';
import { HttpStart } from 'kibana/public';
import { displayArray, displayBoolean, displayObject, renderTextFlexItem } from './display-utils';
import { API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import {
  AuditLoggingSettings,
  ComplianceSettings,
  GeneralSettings,
} from '../panels/audit-logging/types';

export function renderStatusPanel(onSwitchChange: () => void, auditLoggingEnabled: boolean) {
  return (
    <EuiPanel>
      <EuiTitle>
        <h3>Audit logging</h3>
      </EuiTitle>
      <EuiHorizontalRule />
      <EuiForm>
        <EuiDescribedFormGroup
          title={<h3>Enable audit logging</h3>}
          description={<>Enable or disable audit logging</>}
        >
          <EuiFormRow>
            <EuiSwitch
              name="auditLoggingEnabledSwitch"
              label={displayBoolean(auditLoggingEnabled)}
              checked={auditLoggingEnabled}
              onChange={onSwitchChange}
            />
          </EuiFormRow>
        </EuiDescribedFormGroup>
      </EuiForm>
    </EuiPanel>
  );
}

export function renderGeneralSettings(generalSettings: GeneralSettings) {
  return (
    <>
      {renderGeneralSettingsForLayer(generalSettings)}

      <EuiSpacer />

      {renderGeneralSettingsForAttribute(generalSettings)}

      <EuiSpacer />

      {renderGeneralSettingsForIgnore(generalSettings)}
    </>
  );
}

function renderGeneralSettingsForLayer(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Layer settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('REST layer', displayBoolean(generalSettings.enable_rest))}
        {renderTextFlexItem('Transport layer', displayBoolean(generalSettings.enable_transport))}
        {renderTextFlexItem(
          'REST disabled categories',
          displayArray(generalSettings.disabled_rest_categories)
        )}
        {renderTextFlexItem(
          'Transport disabled categories',
          displayArray(generalSettings.disabled_transport_categories)
        )}
      </EuiFlexGrid>
    </>
  );
}

function renderGeneralSettingsForAttribute(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Attribute settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('Bulk requests', displayBoolean(generalSettings.resolve_bulk_requests))}
        {renderTextFlexItem('Resolve indices', displayBoolean(generalSettings.resolve_indices))}
        {renderTextFlexItem('Request body', displayBoolean(generalSettings.log_request_body))}
        {renderTextFlexItem(
          'Sensitive headers',
          displayBoolean(generalSettings.exclude_sensitive_headers)
        )}
      </EuiFlexGrid>
    </>
  );
}

function renderGeneralSettingsForIgnore(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Ignore settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('Exclude users', displayArray(generalSettings.ignore_users))}
        {renderTextFlexItem('Exclude requests', displayArray(generalSettings.ignore_requests))}
      </EuiFlexGrid>
    </>
  );
}

export function renderComplianceSettings(complianceSettings: ComplianceSettings) {
  return (
    <>
      {renderTextFlexItem('Compliance mode', displayBoolean(complianceSettings.enabled))}

      <EuiSpacer />

      <EuiPanel>
        <EuiTitle>
          <h4>Read</h4>
        </EuiTitle>

        <EuiSpacer />

        <EuiFlexGrid columns={3}>
          {renderTextFlexItem(
            'Read metadata only',
            displayBoolean(complianceSettings.read_metadata_only)
          )}
          {renderTextFlexItem('Ignored users', displayArray(complianceSettings.read_ignore_users))}
          {/* read_watched_fields is object instead of array.*/}
          {renderTextFlexItem(
            'Watched fields',
            displayObject(complianceSettings.read_watched_fields)
          )}
        </EuiFlexGrid>
      </EuiPanel>

      <EuiSpacer />

      <EuiPanel>
        <EuiTitle>
          <h4>Write</h4>
        </EuiTitle>

        <EuiSpacer />

        <EuiFlexGrid columns={3}>
          {renderTextFlexItem(
            'Write metadata only',
            displayBoolean(complianceSettings.write_metadata_only)
          )}
          {renderTextFlexItem('Log diffs', displayBoolean(complianceSettings.write_log_diffs))}
          {renderTextFlexItem('Ignored users', displayArray(complianceSettings.write_ignore_users))}
          {renderTextFlexItem(
            'Watched indices',
            displayArray(complianceSettings.write_watched_indices)
          )}
        </EuiFlexGrid>
      </EuiPanel>
    </>
  );
}

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await http.post(`${API_ENDPOINT_AUDITLOGGING_UPDATE}`, {
    body: JSON.stringify(updateObject),
  });
}
