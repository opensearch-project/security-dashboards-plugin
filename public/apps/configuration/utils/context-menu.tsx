/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiButtonProps,
  EuiPopover,
  EuiContextMenuPanel,
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
} from '@elastic/eui';
import { useState } from 'react';
import React from 'react';

export function useContextMenuState(
  buttonText: string,
  buttonProps: EuiButtonProps,
  children: React.ReactElement[]
): [React.ReactElement, () => void] {
  const [isContextMenuOpen, setContextMenuOpen] = useState<boolean>(false);
  const closeContextMenu = () => setContextMenuOpen(false);

  const button = (
    <EuiButton
      iconType="arrowDown"
      iconSide="right"
      onClick={() => {
        setContextMenuOpen(true);
      }}
      {...buttonProps}
    >
      {buttonText}
    </EuiButton>
  );

  const items = [
    <EuiFlexGroup direction="column" alignItems="flexStart" gutterSize="s">
      {children.map((child, index) => (
        <EuiFlexItem key={index}>{child}</EuiFlexItem>
      ))}
    </EuiFlexGroup>,
  ];

  const component = (
    <EuiPopover
      id={buttonText}
      button={button}
      isOpen={isContextMenuOpen}
      closePopover={closeContextMenu}
      panelPaddingSize="s"
    >
      <EuiContextMenuPanel items={items} hasFocus={false} />
    </EuiPopover>
  );

  return [component, closeContextMenu];
}
