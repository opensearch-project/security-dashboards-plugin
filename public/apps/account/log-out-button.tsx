/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { EuiButtonEmpty } from '@elastic/eui';
import { HttpStart } from 'opensearch-dashboards/public';
import { externalLogout, logout } from './utils';
import { AuthType, OPENID_AUTH_LOGOUT, SAML_AUTH_LOGOUT } from '../../../common';
import { setShouldShowTenantPopup } from '../../utils/storage-utils';

export function LogoutButton(props: {
  authType: string;
  http: HttpStart;
  divider: JSX.Element;
  logoutUrl?: string;
}) {
  if (props.authType === AuthType.OPEN_ID) {
    return (
      <div>
        {props.divider}
        <EuiButtonEmpty
          data-test-subj="log-out-2"
          color="danger"
          size="xs"
          onClick={() => externalLogout(props.http, OPENID_AUTH_LOGOUT)}
        >
          Log out
        </EuiButtonEmpty>
      </div>
    );
  } else if (props.authType === AuthType.SAML) {
    return (
      <div>
        {props.divider}
        <EuiButtonEmpty
          data-test-subj="log-out-1"
          color="danger"
          size="xs"
          onClick={() => externalLogout(props.http, SAML_AUTH_LOGOUT)}
        >
          Log out
        </EuiButtonEmpty>
      </div>
    );
  } else if (props.authType === AuthType.PROXY) {
    setShouldShowTenantPopup(null);
    return <div />;
  } else {
    return (
      <div>
        {props.divider}
        <EuiButtonEmpty
          data-test-subj="log-out-3"
          color="danger"
          size="xs"
          onClick={() => logout(props.http, props.logoutUrl)}
        >
          Log out
        </EuiButtonEmpty>
      </div>
    );
  }
}
