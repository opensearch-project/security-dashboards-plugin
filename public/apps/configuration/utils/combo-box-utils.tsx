/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';
import { Dispatch, SetStateAction } from 'react';
import { PropertyName, curry } from 'lodash';
import { appendElementToArray } from './array-state-utils';

// Build an option object for EuiComboBox
export function stringToComboBoxOption(option: string): EuiComboBoxOptionOption {
  return { label: option };
}

// Unbuild an EuiComboBox option objects to string
export function comboBoxOptionToString(option: EuiComboBoxOptionOption): string {
  return option.label;
}

/**
 * Convert a string to a combo box option and append to an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param newValue value to be updated
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [{label: 1}, {label: 2}]
 *  path = []
 *  newValue = 3
 * The newState will be [{label: 1}, {label: 2}, {label: 3}]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [
 *    {options: [{label: 1}, {label: 2}]},
 *    {options: [{label: 3}, {label: 4}]}
 *  ]
 *  path = [0, 'options']
 *  newValue = 5
 * The newState will be [
 *    {options: [{label: 1}, {label: 2}, {label: 5}]},
 *    {options: [{label: 3}, {label: 4}]}
 * ]
 */
export function appendOptionToComboBox(
  setStateFunc: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: string
) {
  appendElementToArray(setStateFunc, path, { label: newValue });
}

export const appendOptionToComboBoxHandler = curry(appendOptionToComboBox);
