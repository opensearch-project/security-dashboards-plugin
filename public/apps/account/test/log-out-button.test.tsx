/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import { LogoutButton } from '../log-out-button';
import { logout } from '../utils';
import { AuthType } from '../../../../common';

jest.mock('../utils', () => ({
  logout: jest.fn(),
}));

describe('Account menu - Log out button', () => {
  const mockHttpStart = {
    basePath: {
      serverBasePath: '',
    },
  };
  const mockDivider = <></>;
  describe('renders', () => {
    it('renders when auth type is MultiAuth: openid', () => {
      const component = shallow(
        <LogoutButton authType={AuthType.OPEN_ID} http={mockHttpStart} divider={mockDivider} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders when auth type is MultiAuth: saml', () => {
      const component = shallow(
        <LogoutButton authType={AuthType.SAML} http={mockHttpStart} divider={mockDivider} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders when auth type is MultiAuth: basicauth', () => {
      const component = shallow(
        <LogoutButton authType={AuthType.BASIC} http={mockHttpStart} divider={mockDivider} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders when auth type is OpenId', () => {
      const component = shallow(
        <LogoutButton authType={AuthType.OPEN_ID} http={mockHttpStart} divider={mockDivider} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders when auth type is SAML', () => {
      const component = shallow(
        <LogoutButton authType={AuthType.SAML} http={mockHttpStart} divider={mockDivider} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders when auth type is Proxy', () => {
      const component = shallow(
        <LogoutButton authType={AuthType.PROXY} http={mockHttpStart} divider={mockDivider} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders when auth type is not OpenId, SAML, or Proxy', () => {
      const component = shallow(
        <LogoutButton authType="dummy" http={mockHttpStart} divider={mockDivider} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  it('should call logout function on click when auth type is not OpenId, SAML, or Proxy', () => {
    const component = shallow(
      <LogoutButton authType="dummy" http={mockHttpStart} divider={mockDivider} />
    );
    component.find('[data-test-subj="log-out-3"]').simulate('click');

    expect(logout).toBeCalled();
  });
});
