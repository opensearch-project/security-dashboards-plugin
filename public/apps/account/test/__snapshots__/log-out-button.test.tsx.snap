// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account menu - Log out button renders renders when auth type is MultiAuth: basicauth 1`] = `
<div>
  <EuiButtonEmpty
    color="danger"
    data-test-subj="log-out-3"
    onClick={[Function]}
    size="xs"
  >
    Log out
  </EuiButtonEmpty>
</div>
`;

exports[`Account menu - Log out button renders renders when auth type is MultiAuth: openid 1`] = `
<div>
  <EuiButtonEmpty
    color="danger"
    data-test-subj="log-out-2"
    onClick={[Function]}
    size="xs"
  >
    Log out
  </EuiButtonEmpty>
</div>
`;

exports[`Account menu - Log out button renders renders when auth type is MultiAuth: saml 1`] = `
<div>
  <EuiButtonEmpty
    color="danger"
    data-test-subj="log-out-1"
    onClick={[Function]}
    size="xs"
  >
    Log out
  </EuiButtonEmpty>
</div>
`;

exports[`Account menu - Log out button renders renders when auth type is OpenId 1`] = `
<div>
  <EuiButtonEmpty
    color="danger"
    data-test-subj="log-out-2"
    onClick={[Function]}
    size="xs"
  >
    Log out
  </EuiButtonEmpty>
</div>
`;

exports[`Account menu - Log out button renders renders when auth type is Proxy 1`] = `<div />`;

exports[`Account menu - Log out button renders renders when auth type is SAML 1`] = `
<div>
  <EuiButtonEmpty
    color="danger"
    data-test-subj="log-out-1"
    onClick={[Function]}
    size="xs"
  >
    Log out
  </EuiButtonEmpty>
</div>
`;

exports[`Account menu - Log out button renders renders when auth type is not OpenId, SAML, or Proxy 1`] = `
<div>
  <EuiButtonEmpty
    color="danger"
    data-test-subj="log-out-3"
    onClick={[Function]}
    size="xs"
  >
    Log out
  </EuiButtonEmpty>
</div>
`;
